kuma/dashboards/utils.py

```diff
@@ -0,0 +1,281 @@
+        created__range=(day, next_day)).only('review')
+    for blocked in blocked_edits:
+        # Is it a false positive?
+        if blocked.review == DocumentSpamAttempt.HAM:
+            counts['blocked_ham'] += 1
+        elif blocked.review == DocumentSpamAttempt.SPAM:
+            counts['blocked_spam'] += 1
+        else:
+            if blocked.review == DocumentSpamAttempt.NEEDS_REVIEW:
+                needs_review = True
+            continue
+
+    events = {
+        'published_spam': 0,
+        'published_ham': 0,
+        'blocked_spam': 0,
+        'blocked_ham': 0,
+    }
+    events.update(counts)
+
+    return {
+        'version': 1,
+        'generated': datetime.datetime.now().isoformat(),
+        'day': day.isoformat(),
+        'needs_review': needs_review,
+        'events': events
+    }
+
+
+def spam_dashboard_historical_stats(periods=None, end_date=None):
+    """
+    Gather spam statistics for a range of dates.
+
+    Keywords Arguments:
+    periods - a sequence of (days, identifier, name) tuples
+    end_date - The ending anchor date for the statistics
+    """
+    from .jobs import SpamDayStats, SpamDashboardRecentEvents
+
+    periods = periods or SPAM_PERIODS
+    if hasattr(end_date, 'date'):
+        end_date = end_date.date()
+    end_date = end_date or (datetime.date.today() - datetime.timedelta(days=1))
+
+    longest = max(days for days, identifier, name in periods)
+    spans = [
+        (identifier,
+         days,
+         name,
+         end_date - datetime.timedelta(days=days - 1),  # current period begins
+         end_date)
+        for days, identifier, name in periods
+    ]
+
+    start_date = end_date - datetime.timedelta(days=longest - 1)
+    dates = date_range(start_date, end_date)
+
+    # Iterate over the daily stats
+    job = SpamDayStats()
+    trends = defaultdict(Counter)
+    for day in dates:
+        # Gather daily raw stats
+        raw_events = job.get(day)
+        day_events = raw_events['events']
+
+        # Regenerate stats if there are change attempts with
+        # needs_review marked and the data is stale.
+        if raw_events['needs_review']:
+            generated = parser.parse(raw_events['generated'])
+
+            age = datetime.datetime.now() - generated
+            if age.total_seconds() > 300:
+                job.invalidate(day)
+
+        # Accumulate trends over periods
+        for period_id, length, period_name, start, end in spans:
+            # Sum up the items in day_events with any items that may
+            # already be in the Counter at trends[period_id]
+            if start <= day <= end:
+                trends[period_id].update(day_events)
+
+    # Prepare output data
+    data = {
+        'version': 1,
+        'generated': datetime.datetime.now().isoformat(),
+        'end_date': end_date,
+        'periods': periods,
+        'trends': [],
+    }
+
+    job = SpamDashboardRecentEvents()
```

Comment:
The time that data is available bugs me, and it feels like the caches are wrong, but I don't have a solution yet.

* We know a published revision is spam as soon as a moderator marks it as spam
* The page views of a spam revision aren't known until 24 - 48 hours after the revision is reverted, due to data processing delays in Google Analytics.
* Once a revision is reverted, there should be no new page views, so you shouldn't have to ask again.  It is hard to determine if GA is done counting, though.
* Google Analytics uses sampling to determine the number of unique page views, and can change 10% - 20% from day to day, if you change the date range, etc.
* We need to know the page views for a revision for the long term stats, but we won't know the numbers for recent spam for 24-48 hours.

This suggests to me that we need:
* Long-term storage for revision page views that doesn't kick in until the revision is old enough
* Short-term storage or no storage (i.e. on-demand queries) for published spam
* Some storage for the collected statistics (day, week, month) for dashboard speed.

I'm inclined to drop all the job-level caching, and just use a 30-60 minute cache on the spam dashboard page itself, until we can figure out a better way to gather and store the page view data.